webauthn: Client support behind a config flag

jblz opened the pull request:
This pull request adds front-end support for using hardware tokens / webauthn as a second factor. 

### Changes proposed in this Pull Request

* Add a new section in `/me/security/two-step` where a user with a browser that supports webauthn can add new keys.
* Adds a new login block/action that handles the authentication part (second factor) when a given account has a registered hardware device.

### Testing instructions

Test live: https://calypso.live/?branch=add/gm19-webauthn
To add a new security key, one needs to setup first the standard two factor authentication. For now, this requires a proxied connection.

- Navigate to `/me/security/two-step`.
- Under the `Security Key` section, register a new key. Follow the instructions.
- Under a private window, log in to your test account. Make sure you are prompted to use the registered key. Click on the `Continue with security key` button and follow the instructions.

### Screenshots

#### Register key section
Before
<img width="733" alt="Screenshot 2019-09-20 at 16 12 51" src="https://user-images.githubusercontent.com/1081870/65334655-1fcac980-dbc3-11e9-868a-2a9ebd4add0a.png">
After
<img width="730" alt="Screenshot 2019-09-20 at 16 13 20" src="https://user-images.githubusercontent.com/1081870/65334661-25c0aa80-dbc3-11e9-9633-5839e7b2fbdd.png">

#### Login flow with a registered key
<img width="436" alt="Screenshot 2019-09-20 at 16 19 21" src="https://user-images.githubusercontent.com/1081870/65334682-32dd9980-dbc3-11e9-9021-8b23a6f0af61.png">

belcherj committed:
Add initial basic UI

belcherj committed:
Rename 2fa Key to U2f Key

belcherj committed:
Added more UI

belcherj committed:
Added more UI

belcherj committed:
Move files into main u2f folder

belcherj committed:
Move files into main u2f folder

belcherj committed:
Add u2f code

belcherj committed:
Add u2f code

belcherj committed:
Trigger registration on button

belcherj committed:
Trigger registration on button

bisko committed:
Add delete button and update list styling

bisko committed:
Add delete button and update list styling

bisko committed:
Improve styling a bit

bisko committed:
Improve styling a bit

bisko committed:
Add confirmation popup when deleting a key

bisko committed:
Add confirmation popup when deleting a key

belcherj committed:
Make ket registration work

belcherj committed:
Make ket registration work

bisko committed:
Improvements to data structure and rendering

bisko committed:
Improvements to data structure and rendering

bisko committed:
Update styling, remove row highlight, align delete button

bisko committed:
Update styling, remove row highlight, align delete button

bisko committed:
Update styling to resemble application passwords section

bisko committed:
Update styling to resemble application passwords section

bisko committed:
Improve handling of adding key screen and hide the key list when a key is being added

bisko committed:
Improve handling of adding key screen and hide the key list when a key is being added

bisko committed:
Improve add key screen styling

bisko committed:
Improve add key screen styling

bisko committed:
Fix additional whitespace between items in security keys list

bisko committed:
Fix additional whitespace between items in security keys list

belcherj committed:
Update css

belcherj committed:
Update css

belcherj committed:
Add unsupported text

belcherj committed:
Add unsupported text

belcherj committed:
Move register command to componentDidMount

belcherj committed:
Move register command to componentDidMount

bisko committed:
Merge things

bisko committed:
Fix CSS

bisko committed:
Fix CSS

bisko committed:
More CSS fixes

bisko committed:
More CSS fixes

bisko committed:
More CSS fixes

bisko committed:
More CSS fixes

belcherj committed:
Add api enpoints

belcherj committed:
Add api enpoints

Commit:
Fix check for non-supported browsers

Commit:
Fix check for non-supported browsers

belcherj committed:
Use api registration request

belcherj committed:
Use api registration request

belcherj committed:
Add delete

belcherj committed:
Add delete

bisko committed:
Add config flags to enable u2f key management only in dev and staging

bisko committed:
Add config flags to enable u2f key management only in dev and staging

belcherj committed:
Make apis work

belcherj committed:
Make apis work

xknown committed:
Stash

xknown committed:
Stash

xknown committed:
Stash

xknown committed:
Stash

xknown committed:
Make the remove button work

xknown committed:
Make the remove button work

xyu committed:
Merge remote-tracking branch 'xknown/add/manage-u2f-key' into add/gm19-webauthn

xyu committed:
Update npm shrinkwrap

xyu committed:
Remove leftover merge files

xyu committed:
Remove leftover merge files

xyu committed:
Update shrinkwrap

```
$ npm run update-deps
```

xyu committed:
Fix login page imports

xyu committed:
Update Gridicon

xyu committed:
Fix more merge errors

- Use muriel colors
- Fix imports for 2fa jsx

xyu committed:
Fix more merge errors

- Use muriel colors
- Fix imports for 2fa jsx

xyu committed:
Remove old U2F code

xyu committed:
Remove old U2F code

xyu committed:
Revert npm-shrinkwrap.json

xknown committed:
Fix the demo

xknown committed:
Fix the demo

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xyu committed:
Remove feature flag ui/first-view

xknown committed:
Remove login related stuff, it breaks things

xknown committed:
Remove login related stuff, it breaks things

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xyu committed:
Remove unreferenced file

xyu committed:
Remove unreferenced file

xyu committed:
Remove references to u2f the old standard

xyu committed:
Remove references to u2f the old standard

a8ci18n commented on client/me/security-2fa-key/delete-item-button.jsx:
```diff
@@ -0,0 +1,90 @@
+			{ action: 'cancel', label: this.props.translate( 'Cancel' ) },
+			{
+				action: 'delete',
+				label: this.props.translate( 'Remove key' ),
```
Hi! I've found a possible matching string that has already been translated **14** times:
`translate( 'Remove Key' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/me/security-2fa-key/delete-item-button.jsx:
```diff
@@ -0,0 +1,90 @@
+					buttons={this.buttons}
+					onClose={this.onCloseDialog}
+				>
+					<h1>{this.props.translate( 'Remove key?' )}</h1>
```
Hi! I've found a possible matching string that has already been translated **14** times:
`translate( 'Remove Key' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/me/security-2fa-key/index.jsx:
```diff
@@ -0,0 +1,150 @@
+		const { addingKey, isBrowserSupported, errorMessage, security2faKeys } = this.state;
+		return (
+			<Fragment>
+				<SectionHeader label={ translate( 'Security Key' ) }>
```
Hi! I've found a possible matching string that has already been translated **25** times:
`translate( 'Security Key Added' )` *ES Score: 6*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/gm19-webauthn)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

xyu commented on client/lib/webauthn/index.js:
```diff
@@ -0,0 +1,141 @@
+const POST = 'POST';
+
+function strToBin(str) {
+	str = str.replace(/[-_]/g, function(m) { return m[0] === '-' ? '+' : '/'; });
```
This looks a bit fishy, we're changing the string before converting it to binary but below in `binToStr()` we're not reversing the change. Does this conversion happen in PHP? Is this just broken?

xknown replied:
The values that come from the API side are encoded using with the base64url, which `atob` doesn't handle. That's why it's converted to the standard `base64` encoding.
We can potentially add the same conversion to binToStr to avoid this confusion -- the PHP version handles the `base64`, and `base64url` encodings.

jblz committed:
Code style & lint fixes:
* Format webauthn/index.js with prettier
* remove unused vars
* let instead of var

jblz committed:
Code style & lint fixes:
* Format webauthn/index.js with prettier
* remove unused vars
* let instead of var

jblz commented on config/development.json:
```diff
@@ -4,7 +4,7 @@
 	"client_slug": "browser",
 	"directly_rtm_widget_environment": "sandbox",
-	"protocol": "http",
+	"protocol": "https",
```
This should be in the test instructions instead of the PR

xknown replied:
I agree, I was planning to revert this change.

jblz committed:
pull out @format annotations & lint fixes

jblz committed:
pull out @format annotations & lint fixes

xknown committed:
Revert the accidental protocol change

xknown committed:
Revert the accidental protocol change

jblz committed:
guard response reference with lodash get

jblz committed:
guard response reference with lodash get

jblz committed:
mv webauthn declaration to "internal" block

jblz committed:
mv webauthn declaration to "internal" block

jblz committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

jblz committed:
Fix Gridicon import & prettier format

jblz committed:
Fix Gridicon import & prettier format

bluefuton commented on client/me/security-2fa-key/add.jsx:
```diff
@@ -0,0 +1,66 @@
+	render() {
+		return (
+			<Card>
+				<Fragment>
```
Don't think we need `<Fragment>` here 👍

xyu replied:
Removed in c05f15d

bluefuton commented on client/me/security-2fa-key/style.scss:
```diff
@@ -0,0 +1,43 @@
+		flex-shrink: 0;
+
+		.security-2fa-key__item-subtitle {
+			color: $muriel-gray-400;
```
Nit: use `color: var( --muriel-gray-400 )`.

xyu replied:
Fixed in 811f6bd

bluefuton commented on client/me/security-2fa-key/style.scss:
```diff
@@ -0,0 +1,43 @@
+	display: flex;
+	align-items: center;
+
+	.security-2fa-key__item-information {
```
No need to nest these classes (`.security-2fa-key__item-information`, `.security-2fa-key__item-subtitle` and `.security-2fa-key__delete-key`).

xyu replied:
Fixed in 811f6bd

jblz committed:
Put the gridicon imports back
see: p1568401556121800-slack-C02DQP0FP

jblz committed:
Put the gridicon imports back
see: p1568401556121800-slack-C02DQP0FP

xyu committed:
Cleanup Key Registration Errors

Standardize on returning a `Promise.reject()` object with standardized error codes:

- `DuplicateKey` -- the key being registered is already registered
- `TimeoutCanceled` -- no interaction from user to confirm signing or canceled by user
- `Canceled` -- user canceled allowing key interaction / signing
- `Unknown` / `NoResponse` -- misc errors

xyu committed:
Cleanup Key Registration Errors

Standardize on returning a `Promise.reject()` object with standardized error codes:

- `DuplicateKey` -- the key being registered is already registered
- `TimeoutCanceled` -- no interaction from user to confirm signing or canceled by user
- `Canceled` -- user canceled allowing key interaction / signing
- `Unknown` / `NoResponse` -- misc errors

david-binda committed:
Error displaying

This commit takes advantage of error handling in webauthn.register to display the messages to users.

david-binda committed:
Error displaying

This commit takes advantage of error handling in webauthn.register to display the messages to users.

xyu committed:
Example sending key name when registering

xyu committed:
Example sending key name when registering

xyu commented:
7a9abff shows how to send a key name which works with D32804-code

xyu commented on client/me/security-2fa-key/add.jsx:
```diff
@@ -23,19 +25,24 @@ class Security2faKeyAdd extends React.Component {
 				this.keyRegistered();
 			} )
 			.catch( err => {
-				this.props.onCancel( err );
```
@david-binda don't we need to still call the onCancel func so that we can call `addKeyCancel` from index.jsx?

xyu replied:
Never mind I get it it's doing `onCancel` when the message is dismissed! Awesome!

xyu commented on client/me/security-2fa-key/add.jsx:
```diff
@@ -23,19 +25,24 @@ class Security2faKeyAdd extends React.Component {
 			} )
 			.catch( err => {
-				this.props.onCancel( err );
+				this.setState( { error: err } );
```
Nit pick maybe we can just name `err` as `error` above so that we can just do:

```
this.setState( { error } );
```

I think this works in ES6?

david-binda replied:
Fixed in f41e605 . Thanks for the tip, I had no idea this can be done!

xyu committed:
Cleanup Styles Props @bluefuton

xyu committed:
Cleanup Styles Props @bluefuton

xyu committed:
Remove Fragment Props @bluefuton

xyu committed:
Remove Fragment Props @bluefuton

david-binda committed:
Rename the err to error in the catch statement

Allows us to simplify the setState call, per @xyu 's feedback. Thanks!

david-binda committed:
Rename the err to error in the catch statement

Allows us to simplify the setState call, per @xyu 's feedback. Thanks!

david-binda committed:
Make the error messages translatable

david-binda committed:
Make the error messages translatable

xyu committed:
Show key names and shorten default name

xyu committed:
Show key names and shorten default name

jblz committed:
suppress eslint error on the Gridicon import
...until the fix lands

jblz committed:
suppress eslint error on the Gridicon import
...until the fix lands

jblz committed:
Add/gm19 webauthn -- challenge & response pieces (#36134)

* Add the authenticate part

* Add the security-key-form component

* Add support to webauthn authentication

* Move the webauthn authentication to the appropriate place.

* Export utility functions

* Handle errors and update the nonce on failed requests.

* Update the data attribute to avoid (potentially) breaking e2e tests.

* Cleanup some lint errors and use SecurityKey instead of HardwareKey

* Remove unused code

jblz committed:
Add/gm19 webauthn -- challenge & response pieces (#36134)

* Add the authenticate part

* Add the security-key-form component

* Add support to webauthn authentication

* Move the webauthn authentication to the appropriate place.

* Export utility functions

* Handle errors and update the nonce on failed requests.

* Update the data attribute to avoid (potentially) breaking e2e tests.

* Cleanup some lint errors and use SecurityKey instead of HardwareKey

* Remove unused code

jblz committed:
remove stray `n` property from the Security Key Button

jblz committed:
remove stray `n` property from the Security Key Button

jblz committed:
lint fixes

jblz committed:
lint fixes

xyu committed:
Allow calypso.live to use keys

xyu committed:
Allow calypso.live to use keys

xyu committed:
Fix language, props @megsfulton

xyu committed:
Fix language, props @megsfulton

xyu committed:
Send origin from calypso.live

xyu committed:
Send origin from calypso.live

xyu committed:
Send host from everywhere but prod

xyu committed:
Send host from everywhere but prod

xyu committed:
Use env_id config to see enviroment

xyu committed:
Use env_id config to see enviroment

david-binda committed:
Initial version of the security key name form

This implements the basics for allowing the security keys to be named before they are added.

david-binda committed:
Initial version of the security key name form

This implements the basics for allowing the security keys to be named before they are added.

david-binda committed:
Use the state/notices/actions for error handling

david-binda committed:
Use the state/notices/actions for error handling

david-binda committed:
Remove the cancel button

The cancel button is provided by browsers and we would need to make the promise we are using cancellable in order to not break the React flow anyway. Makes things easier.

david-binda committed:
Remove the cancel button

The cancel button is provided by browsers and we would need to make the promise we are using cancellable in order to not break the React flow anyway. Makes things easier.

david-binda committed:
Make the Warning notice to disappear after 5s

david-binda committed:
Make the Warning notice to disappear after 5s

xyu committed:
Move Security Key Section Up

xyu committed:
Move Security Key Section Up

xyu committed:
Cleanup lint errors and disallow blank key names

xyu committed:
Cleanup lint errors and disallow blank key names

xknown committed:
Request the security credential without user interaction

This removes the need of using the `Continue` button.

xknown committed:
Request the security credential without user interaction

This removes the need of using the `Continue` button.

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xknown committed:
Revert an accidental change forcing webauthn as the first option

xknown committed:
Revert an accidental change forcing webauthn as the first option

xyu committed:
Cleanup more lint errors

xyu committed:
Cleanup more lint errors

jblz committed:
connect notice actions to dispatch

jblz committed:
connect notice actions to dispatch

jblz committed:
export individual functions from webauthn lib (#36140)

* export individual functions from webauthn lib
* Rename some functions for clarity

jblz committed:
export individual functions from webauthn lib (#36140)

* export individual functions from webauthn lib
* Rename some functions for clarity

david-binda committed:
Add success notice when key is added.

david-binda committed:
Add success notice when key is added.

david-binda committed:
Update the success message wording

david-binda committed:
Update the success message wording

david-binda committed:
Removing "Read more" part of the Safari unsupported message

david-binda committed:
Removing "Read more" part of the Safari unsupported message

david-binda committed:
Adjust the messaging for browsers not supporting the FIDO2 standard.

david-binda committed:
Adjust the messaging for browsers not supporting the FIDO2 standard.

david-binda committed:
Improving the waiting screen messaging

david-binda committed:
Improving the waiting screen messaging

xknown committed:
Revert 86978a3

xknown committed:
Revert 86978a3

jblz committed:
webauthn: Rename all instances of u2f to webauthn (#36156)

* webauthn: Rename all instances of u2f to webauthn

* Rename u2f to webauthn

jblz committed:
webauthn: Rename all instances of u2f to webauthn (#36156)

* webauthn: Rename all instances of u2f to webauthn

* Rename u2f to webauthn

jblz committed:
rename css class to standard

jblz committed:
rename css class to standard

sgomes committed:
Extract WaitForKey component.

sgomes committed:
Extract WaitForKey component.

jblz committed:
Pull out unused JSX fragment wrappers & prettier formatting

jblz committed:
Pull out unused JSX fragment wrappers & prettier formatting

jblz committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

jblz committed:
Merge branch 'master' into add/gm19-webauthn

xknown committed:
Do not show the security key prompt for browsers that don't support it

xknown committed:
Do not show the security key prompt for browsers that don't support it

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xknown committed:
Send the nonce to the authenticate endpoint

xknown committed:
Send the nonce to the authenticate endpoint

xknown committed:
Use by default the remember_me setting so the 2fa cookies

xknown committed:
Use by default the remember_me setting so the 2fa cookies

jblz committed:
rm unused componentDidMount function

jblz committed:
rm unused componentDidMount function

jblz committed:
turn off autocomplete on the key name input

jblz committed:
turn off autocomplete on the key name input

jblz committed:
camelCase the autoComplete attribute so React doesn't complain

jblz committed:
camelCase the autoComplete attribute so React doesn't complain

xknown committed:
Add focus to the security key name

This focus the key name element when registering a new key.

xknown committed:
Add focus to the security key name

This focus the key name element when registering a new key.

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xknown committed:
Disable the button when we are waiting for the user

xknown committed:
Disable the button when we are waiting for the user

jblz commented on client/me/security-2fa-key/name.jsx:
```diff
@@ -0,0 +1,74 @@
+	};
+
+	componentDidMount = () => {
+		this.setState( { error: false, keyName: '' }, () => {
```
Do we really need the `setState` call in `componentDidMount`? The state is initialized to the same values in the statement above.

xknown replied:
I'd say it's not needed.

xyu replied:
Kill it with 🔥

xknown committed:
Add a spinner to the logging flow

xknown committed:
Add a spinner to the logging flow

kraftbj committed:
Add periods to all error messages to match the success message.

kraftbj committed:
Add periods to all error messages to match the success message.

xknown committed:
Use the appropriate message on the login flow

xknown committed:
Use the appropriate message on the login flow

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xknown committed:
Remove the unneeded `setState` call

xknown committed:
Remove the unneeded `setState` call

xknown committed:
Remove unreachable code.

xknown committed:
Remove unreachable code.

xknown committed:
Merge branch 'master' into add/gm19-webauthn

xknown committed:
Remove the unneeded setState call

xyu committed:
Feature gate based on if API calls are allowed and turn feature flag on for all but production. Also remove some logging.

xyu committed:
Feature gate based on if API calls are allowed and turn feature flag on for all but production. Also remove some logging.

jblz commented:
@xyu -- the change in de26390, while useful, means that the section "pops in" if support is declared in the server response:

https://cloudup.com/cYmE639ZFLI

I wonder if we can achieve the same effect without the "pop in" somehow.

xyu commented:
> I wonder if we can achieve the same effect without the "pop in" somehow.

Probably... but I have no idea what's the proper Calypso / React way to accomplish that.

bluefuton commented:
> I have no idea what's the proper Calypso / React way to accomplish that

We sometimes show a placeholder while the element loads (like a post loading in Reader), but that would just cause a "pop out" in this case if the browser support isn't there.

I would suggest always showing the 'Security Keys' section regardless of browser support, and include a note in there for unsupported browsers ("security keys are only supported on x, y and z").

xyu commented:
> I would suggest always showing the 'Security Keys' section regardless of browser support, and include a note in there for unsupported browsers ("security keys are only supported on x, y and z").

So we already show an unsupported browsers message, this really is only an issue right now when we are gating on VIP/proxy access because we don't render the box unless the API call for WebAuthn to the wpcom APIs allow it. I guess how big of a problem this flash / 2 stage load is depends on how fast we turn this on for "everyone" and remove the check.

xknown committed:
Merge branch 'master' into add/gm19-webauthn

xknown committed:
Merge branch 'master' into add/gm19-webauthn

xyu committed:
UX tweaks

- Autofocus on "Continue with security key" button so that we can do {username} -> [enter] -> {password} -> [enter] -> [enter] -> [press key] instead of having to click on it with a mouse.
- Don't hide the spinner after a successful auth so that we go from the "Waiting for security key" card directly to logged in page rather then from "Waiting for security key" back to "Continue with security key" then to logged in page

xyu committed:
UX tweaks

- Autofocus on "Continue with security key" button so that we can do {username} -> [enter] -> {password} -> [enter] -> [enter] -> [press key] instead of having to click on it with a mouse.
- Don't hide the spinner after a successful auth so that we go from the "Waiting for security key" card directly to logged in page rather then from "Waiting for security key" back to "Continue with security key" then to logged in page

xyu commented:
@bluefuton @jblz what's the process for getting things reviewed and merged for Calypso? Should we ping anyone in particular with this change? What should our next steps be?

bluefuton committed:
Minor cleanup

bluefuton committed:
Minor cleanup

bluefuton commented:
A few minor things I spotted, but these can be cleaned up in followup PRs:
- add a 'what's this?' message to explain what a security key is and how to get one
- improve the error message when a security key registration fails (currently just 'security key registration error')
- replace the gridicon imports with `components/gridicon` - the issue we ran into during the GM has since been solved by @sirreal 

<img width="351" alt="Screen Shot 2019-09-25 at 14 57 28" src="https://user-images.githubusercontent.com/17325/65565857-49aa3480-dfa5-11e9-82dc-a5381b6d6c76.png">

bluefuton reviewed (approved):
🚢 Looks good!

sirreal commented on client/lib/webauthn/index.js:
```diff
@@ -0,0 +1,186 @@
+
+export function binToStr( bin ) {
+	return btoa( new Uint8Array( bin ).reduce( ( s, byte ) => s + String.fromCharCode( byte ), '' ) );
+}
```
This pair of functions seems very error prone, and some quick testing indicates that it has some issues. Are they expected to operate on a very limited set of inputs?

Drop this in a browser:

```js
function strToBin( str ) {
	str = str.replace( /[-_]/g, function( m ) {
		return m[ 0 ] === '-' ? '+' : '/';
	} );
	return Uint8Array.from( atob( str ), c => c.charCodeAt( 0 ) );
}
function binToStr( bin ) {
	return btoa( new Uint8Array( bin ).reduce( ( s, byte ) => s + String.fromCharCode( byte ), '' ) );
}

const inputs = [
    [ 'hex', '0123456789abcdef' ],
    [ 'simple', 'Hello, world!' ],
    [ 'complex', 'áæñü?+-_ ?=&' ],
    [ 'emoji', '😱😎\u00a0' ],
];

inputs.forEach( ( [ desc, input ] ) => {
  try {
    if ( input !== binToStr( strToBin( input ) ) ) {
      console.warn( '%o: %o did not match its conversion %o', desc, input, binToStr( strToBin( input ) ) )
    }
  } catch ( err ) {
    console.warn( 'Error in input %o: %o\n%o', desc, input, err.message );
  }
} )
// Error in input "simple": "Hello, world!"
// "Failed to execute 'atob' on 'Window': The string to be decoded is not correctly encoded."
// Error in input "complex": "áæñü?+-_ ?=&"
// "Failed to execute 'atob' on 'Window': The string to be decoded is not correctly encoded."
// Error in input "emoji": "😱😎 "
// "Failed to execute 'atob' on 'Window': The string to be decoded contains characters outside of the Latin1 range."
```

Should we add documentation, unit tests, and/or fixes and improvements? Can we use an external library?

xknown replied:
Thanks for the review! I think the function names should be probably renamed to `base64StringToBin/binStringToBase64` or something similar, `strToBin` is only expected to work with base64url encoded strings. We can likely use `base64-js` for `binToStr`.

xknown replied:
Looks like we can use https://www.npmjs.com/package/@github/webauthn-json, will do some tests and see if we can replace the current library.

xyu replied:
That sounds like a great idea @xknown do you want to try and replace the lib in this PR or should we try to do that in a subsequent PR?

xknown replied:
@xyu can you take a look at #36328? I think everything is working as expected, if you notice a problem we can likely postpone merging it to this PR.

xyu replied:
❤️ deleting code
```
+20 −160
```

sirreal commented on client/me/security-2fa-key/delete-item-button.jsx:
```diff
@@ -0,0 +1,87 @@
+import { successNotice } from '../../state/notices/actions';
+import { recordGoogleEvent } from '../../state/analytics/actions';
+import Dialog from 'components/dialog';
+import Gridicon from 'gridicons'; // eslint-disable-line no-restricted-imports
```
```suggestion
import Gridicon from 'components/gridicon';
```

sirreal commented on client/me/security-2fa-key/item.jsx:
```diff
@@ -0,0 +1,46 @@
+}
+
+export default connect(
+	state => ( { state } ),
```
Does this attach the entire application state? Is that necessary?

sirreal commented on client/me/security-2fa-key/list.jsx:
```diff
@@ -0,0 +1,46 @@
+}
+
+export default connect(
+	state => ( { state } ),
```
Does this attach the entire application state? Is that necessary?

sirreal commented on client/me/security-2fa-key/wait-for-key.jsx:
```diff
@@ -0,0 +1,24 @@
+import { useTranslate } from 'i18n-calypso';
+
+export default function WaitForKey() {
+	const translate = useTranslate();
```
🥇  Love the function component 👍

sirreal commented on client/me/security-2fa-key/index.jsx:
```diff
@@ -0,0 +1,158 @@
+/**
+ * External dependencies
+ */
+import Gridicon from 'gridicons'; // eslint-disable-line no-restricted-imports
```
```suggestion
import Gridicon from 'components/gridicon';
```

sirreal commented on client/me/security-2fa-key/index.jsx:
```diff
@@ -0,0 +1,158 @@
+import Notice from 'components/notice';
+
+class Security2faKey extends React.Component {
+	static initialState = Object.freeze( {
```
This doesn't seem to be reused. Is there a reason it's extracted and frozen?

Note that freezing here _does not_ prevent the object inside from being mutated, if that was the intention.

sirreal commented on client/me/security-2fa-key/list.jsx:
```diff
@@ -0,0 +1,46 @@
+ */
+import './style.scss';
+
+class Security2faKeyList extends React.Component {
```
Good candidate for a function component.

sirreal commented on client/me/security-2fa-key/item.jsx:
```diff
@@ -0,0 +1,46 @@
+import { recordGoogleEvent } from '../../state/analytics/actions';
+import Security2faDeleteButton from './delete-item-button';
+
+class Security2faKeyItem extends Component {
```
Good candidate for a function component.

sirreal reviewed (commented):
Thanks for continuing to work on this 🙇 

I've looked the changes over and left some feedback, please consider everything but do not consider it blocking. My only major concern is around the `binToStr` and `strToBin` pair of functions that seem fragile.

Many of my comments are minor or simply observations, so feel free to act on them or not, as you see fit 🙂

xknown committed:
Update client/blocks/login/index.jsx

Remove unnecessary curly brackets

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Update client/blocks/login/index.jsx

Remove unnecessary curly brackets

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Update client/blocks/login/two-factor-authentication/security-key-form.jsx

Remove no longer required auto-format

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Update client/blocks/login/two-factor-authentication/security-key-form.jsx

Remove no longer required auto-format

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xknown committed:
Rename `isDisabled` to `isAuthenticating`

xknown committed:
Rename `isDisabled` to `isAuthenticating`

xknown committed:
Update client/me/security-2fa-key/delete-item-button.jsx

Use `components/gridicon`

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Update client/me/security-2fa-key/delete-item-button.jsx

Use `components/gridicon`

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Update client/me/security-2fa-key/index.jsx

Use `components/gridicon`

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Update client/me/security-2fa-key/index.jsx

Use `components/gridicon`

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

xknown committed:
Remove unused variable

xknown committed:
Remove unused variable

xknown committed:
Convert to function compontents.

xknown committed:
Convert to function compontents.

xknown commented on client/blocks/login/two-factor-authentication/two-factor-actions.jsx:
```diff
@@ -53,20 +54,37 @@ class TwoFactorActions extends Component {
 			<Card className="two-factor-authentication__actions">
+				{ isSecurityKeyAvailable && (
+					<Button data-e2e-link="2fa-security-key-link" onClick={ this.recordSecurityKey }>
+						{ translate( 'Continue with your security\u00A0key' ) }
```
How do you handle the case when the `wordsToKeep` parameter in `preventWidows` is different for each locale? For example, if I use `preventWidows( translate( 'Continue with your security key' ) )`, in french it'd be equivalent to `preventWidows( 'Continuer avec la clé de sécurité' )` which will only make sure that `de sécurité` has the non-breaking space which isn't ideal.

xyu replied:
Do we really need to keep "security key" or "clé de sécurité" together as one word in these labels? I mean it's not really a proper none so maybe having a line break sometimes is ok?

xknown committed:
Rename `strToBin/binToStr

These functions only deal with base64 encoded strings.

xknown committed:
Rename `strToBin/binToStr

These functions only deal with base64 encoded strings.

xknown commented on client/blocks/login/two-factor-authentication/security-key-form.jsx:
```diff
@@ -0,0 +1,103 @@
+		this.props
+			.loginUserWithSecurityKey()
+			.then( () => onSuccess() )
+			.catch( () => this.setState( { isDisabled: false } ) );
```
Yes, I think it's an intended side effect of `onSuccess` (currently it results on a browser redirect). This was recently changed in c6a4b31.

xyu replied:
Yup when we used `finally()` we got a flash of the spinner going away to the click button to start key auth card before getting redirected to logged in view which felt weird / broken so I changed it to only hide the waiting spinner if an auth failure happened so that the user can retry with another key.

xknown committed:
Use arrayBuffer instead of binaryBuffer

xknown committed:
Use arrayBuffer instead of binaryBuffer

xknown committed:
Merge branch 'master' into add/gm19-webauthn

xyu committed:
Skip check of RP ID, WebAuthn allows subdomains of RPs to be valid so defer check to native APIs instead

xyu committed:
Skip check of RP ID, WebAuthn allows subdomains of RPs to be valid so defer check to native APIs instead

xyu committed:
Force new build to test cross subdomain keys on calypso.live

xknown committed:
Add/gm19 webauthn  -- use github webauth json (#36328)

* Use @github/webauthn-json

* Remove all the custom stuff from lib/webauthn

xknown committed:
Add/gm19 webauthn  -- use github webauth json (#36328)

* Use @github/webauthn-json

* Remove all the custom stuff from lib/webauthn

xknown committed:
Merge branch 'master' into add/gm19-webauthn

xknown committed:
Make sure webauthn is only called on a browser context.

xknown committed:
Make sure webauthn is only called on a browser context.

xknown committed:
Merge branch 'master' into add/gm19-webauthn

xknown committed:
Merge branch 'master' into add/gm19-webauthn

jblz committed:
tidy up gridicon import

xknown committed:
Merge branch 'add/gm19-webauthn' of github.com:Automattic/wp-calypso into add/gm19-webauthn

xknown committed:
Transpile `@github/webauthn-json`
